torch/csrc/jit/tracer_state.h

```diff
@@ -59,6 +61,9 @@ struct TracingState : public std::enable_shared_from_this<TracingState> {
   std::shared_ptr<Graph> graph;
   bool active;
 
+  std::unique_ptr<Scope> scope_root;
```

Comment:
I certainly misinterpreted. 
Can you clarify a few things for me:

- In the previous iteration the scopes trie was owned by the root of the graph, as you suggest in 1. We can go back to that design and use a shared pointer for those cases in which a graph is cloned and nodes need to keep pointing to the same scope trie nodes.

- Having pointers in the direction `scopes` -> `ir` requires a greater change, but if we like it more we can go for it. In this case we can't tell a node to store the pointer to the trie node that was current when the node was created. Rather, the scopes trie would have to store pointers to nodes in a set (or vector) at each trie node. This would make Graph scope-free, what do you think about this option?

Thanks for the awesome review BTW

Reply:
I think the second option is nicer overall, but I don't want to stall this PR for any longer, so let's go with 1. for now. We can fix it in a later PR. Is that ok with you?

Reply:
Yep, fine with me. Once this is merged I can work at the refactoring, but probably it's also good to see how #3906 impacts the design